src/mscorlib/src/System/Threading/LazyInitializer.cs

```diff
@@ -241,6 +235,48 @@ private static T EnsureInitializedCore<T>(ref T target, ref bool initialized, re
 
             return target;
         }
+
+        /// <summary>
+        /// Ensure the target is initialized and return the value (slow path). This overload works only for referece type targets.
+        /// Uses the supplied function to create the value.
+        /// </summary>
+        /// <typeparam name="T">The type of target. Has to be reference type.</typeparam>
+        /// <param name="target">A reference to the target to be initialized.</param>
+        /// <param name="syncLock">A reference to a location containing a mutual exclusive lock. If <paramref name="syncLock"/> is null, 
+        /// a new object will be instantiated.</param>
+        /// <param name="valueFactory">
+        /// The <see cref="T:System.Func{T}"/> to invoke in order to produce the lazily-initialized value.
+        /// </param>
+        /// <returns>The initialized object.</returns>
+        private static T EnsureInitializedCore<T>(ref T target, ref object syncLock, Func<T> valueFactory) where T : class
+        {
+            // Lazily initialize the lock if necessary.
+            object slock = syncLock;
+            if (slock == null)
+            {
+                object newLock = new object();
+                slock = Interlocked.CompareExchange(ref syncLock, newLock, null);
+                if (slock == null)
+                {
+                    slock = newLock;
+                }
+            }
```

Comment:
Since this code is now duplicated in two different methods, could you separate it out into a helper? e.g.
```C#
private static object EnsureLockInitialized(ref object syncLock) =>
    syncLock ??
    Interlocked.CompareExchange(ref syncLock, new object(), null) ??
    syncLock;
```
which you could then use here and in the other function as:
```C#
lock (EnsureLockInitialized(ref syncLock))
{
    ...
}
```
?